### PR TITLE
core implementation of efficient last materialization event fetch

### PR DESCRIPTION
**Pull request by prha**
## Summary
This diff adds a method to the event log storage that fetches the last materialization event, which should be stored on the asset key index table (for more recently materialized events).

Adds some tests to capture the backcompat case, which does a more complex query against the event log table.

## Test Plan
BK

### CODE REVIEW

**Comment by prha** [comment]
the internal build will fail until https://github.com/dagster-io/internal/pull/718 is merged

File: python_modules/dagster/dagster/core/storage/event_log/sql_event_log.py [hunk 1]
```diff
@@ -673,6 +673,70 @@ def get_asset_keys(
         asset_keys = [AssetKey.from_db_string(row[0]) for row in sorted(rows, key=lambda x: x[0])]
         return [asset_key for asset_key in asset_keys if asset_key]
 
+    def get_last_materialization_event(
+        self, asset_key_or_keys: Union[AssetKey, List[AssetKey]]
```
Comment (sryza) [inline python_modules/dagster/dagster/core/storage/event_log/sql_event_log.py:677]: Feel free to disregard, but it feels to me like supporting both the single value and the list is overkill here - i.e. it wouldn't be so hard for the caller to just wrap single asset keys in lists, and it saves the `isinstance` step at both the beginning and end of every implementation.

Accepting both types also makes it more difficult to type check the results. Mypy doesn't know that if you pass in an `AssetKey`, you'll get out an `Optional[EventLogEntry]` and not a `Dict[AssetKey, Optional[EventLogEntry]]`.

**Review by sryza — APPROVED** [review]

**Comment by sryza** [comment]
I imagine that, in the near future, we're going to want a version of this that operates at the (asset key, asset partition level)?

File: python_modules/dagster/dagster/core/storage/event_log/base.py [hunk 2]
```diff
@@ -233,6 +233,12 @@ def get_asset_keys(
             asset_keys = asset_keys[:limit]
         return asset_keys
 
+    @abstractmethod
+    def get_last_materialization_event(
```
[the remark this answers is not in the document]
Comment (clairelin135) [reply python_modules/dagster/dagster/core/storage/event_log/base.py:237]: +1

**Review by clairelin135 — APPROVED** [review]

File: python_modules/dagster/dagster/core/storage/event_log/base.py [hunk 3]
```diff
@@ -233,6 +233,12 @@ def get_asset_keys(
             asset_keys = asset_keys[:limit]
         return asset_keys
 
+    @abstractmethod
+    def get_latest_materialization_event(
```
Comment (sryza) [inline python_modules/dagster/dagster/core/storage/event_log/base.py:237]: Should this be plural? Otherwise everything looks great.

**Pull request merged**